File: week9/p4.c
void findDuplicates(int arr[], int n) {
    int *res = (int*)malloc(n * sizeof(int));
    int res_count = 0;

    for (int i = 0; i < n; i++) {
        int index = abs(arr[i]) - 1;
        if (arr[index] < 0) {
            res[res_count++] = index + 1;
        } else {
            arr[index] = -arr[index];
        }
    }

    if (res_count == 0) {
        printf("[]\n");
    } else {
        printf("[");
        for (int i = 0; i < res_count; i++) {
            printf("%d", res[i]);
            if (i != res_count - 1) printf(", ");
        }
        printf("]\n");
    }

    free(res);
}

File: week9/p4.py
class Solution:
    def findDuplicates(self, arr):
        res = []
        for num in arr:
            index = abs(num) - 1
            if arr[index] < 0:
                res.append(index + 1)
            else:
                arr[index] = -arr[index]
        return sorted(res)
